File: decub-gcl/go/consensus.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

// Validator represents a validator
type Validator struct {
	ID   string
	PubKey string
}

// Consensus simulates BFT consensus with quorum signatures
type Consensus struct {
	Validators []Validator
	Threshold  int // >=2/3
}

// NewConsensus creates a new consensus instance
func NewConsensus(validators []Validator) *Consensus {
	threshold := (2 * len(validators)) / 3
	return &Consensus{Validators: validators, Threshold: threshold}
}

// SignBlock simulates signing the block by validators
func (c *Consensus) SignBlock(block Block) []string {
	var signatures []string
	for _, v := range c.Validators {
		// Simulate signature (in real impl, use crypto)
		data := fmt.Sprintf("%s%s", v.ID, HashBlock(block))
		hash := sha256.Sum256([]byte(data))
		sig := hex.EncodeToString(hash[:])
		signatures = append(signatures, sig)
	}
	return signatures
}

// VerifyQuorum checks if there are enough signatures (>=2/3)
func (c *Consensus) VerifyQuorum(signatures []string) bool {
	return len(signatures) >= c.Threshold
}

// ProposeBlock simulates proposing a new block
func (c *Consensus) ProposeBlock(height int, prevHash string, txs []Transaction, proposer string) Block {
	root, _ := BuildMerkleTree(txs)
	header := Header{
		Height:     height,
		PrevHash:   prevHash,
		MerkleRoot: root.Hash,
		Proposer:   proposer,
		Timestamp:  time.Now(),
	}
	return Block{Header: header, Txs: txs}
}

File: decub-gcl/rust/src/consensus.rs
use crate::types::{Block, Header, Transaction, hash_block};
use chrono::Utc;
use sha2::{Digest, Sha256};

#[derive(Clone, Debug)]
pub struct Validator {
    pub id: String,
    pub pub_key: String,
}

#[derive(Clone, Debug)]
pub struct Consensus {
    pub validators: Vec<Validator>,
    pub threshold: usize, // >=2/3
}

impl Consensus {
    pub fn new(validators: Vec<Validator>) -> Self {
        let threshold = (2 * validators.len()) / 3;
        Consensus {
            validators,
            threshold,
        }
    }

    pub fn sign_block(&self, block: &Block) -> Vec<String> {
        self.validators
            .iter()
            .map(|v| {
                let data = format!("{}{}", v.id, hash_block(block));
                let mut hasher = Sha256::new();
                hasher.update(data);
                format!("{:x}", hasher.finalize())
            })
            .collect()
    }

    pub fn verify_quorum(&self, signatures: &[String]) -> bool {
        signatures.len() >= self.threshold
    }

    pub fn propose_block(
        &self,
        height: u64,
        prev_hash: String,
        txs: Vec<Transaction>,
        proposer: String,
    ) -> Block {
        let merkle_root = if let Some((_, root_hash)) = crate::merkle::build_merkle_tree(&txs) {
            root_hash
        } else {
            String::new()
        };
        let header = Header {
            height,
            prev_hash,
            merkle_root,
            proposer,
            timestamp: Utc::now(),
        };
        Block { header, txs }
    }
}
